### PR TITLE
Fix #58326 - Revert selected ranges now behaves like stage/unstage selected ranges

### LINKED ISSUE #58326: Revert selected ranges always reverts all adjecent lines
- VSCode Version: August
- OS Version: Windows 10

Steps to Reproduce:

1. Make a multiline change in file, tracked in git
2. enter diff view
2. Right click -> revert selected range (ctrl + backspace)

Expected: revert only selected range, as in "stage/unstage selected range" behaviouir

Result: vscode always reverts the whole hunk, even though user was selecting specific lines.

Bonus bug: ctrl+z has some erratic behaviour when undoing "revert" action - cursor moves to the bottom of the file for some reason.

### PR DESCRIPTION
**Pull request by Git-Lior**
hey @joaomoreno ,

This fixes #58326 according to the specifications (to be similar to stage/unstage selected). The current behavior looks like this:

revert

As you can see, when reverting a single line of a multiline change, the changed line is reverted, but the entire change is brought back (this is similar to stage/unstage ranges). IMO this is not optimal, but changing that will affect how stage/unstage work as well. What do you think?

### CODE REVIEW

**Comment by joaomoreno** [comment]
@Git-Lior I don't really understand what you've done.

Currently, stage/unstage/revert ranges always talks about **hunks**. Did you just change the behavior to work on **lines** instead?

**Comment by Git-Lior** [comment]
@joaomoreno as I understand it, there are three types of commands:

* `stageAll/unstageAll` - applies to all changes
* `stageChange/revertChange` - applies to a single hunk.
* `stageSelectedRanges/unstageSelectedRanges/revertSelectedRanges` - applies to selected ranges, i.e lines within a hunk (this can be seen in stage/unstage selected implementation).

I made `revertSelectedRanges` implementation consistent with stage/unstage selected. I think this can be improved, but the place to improve it is in `intersectDiffWithRange`, so it would affect stage/unstage selected as well.

**Comment by joaomoreno** [comment]
Currently there are only the first two: all changes and hunks. We do not have line operations yet. We currently call hunks as: `selectedRanges` .

`stageChange/revertChange` and `stageSelectedRanges/unstageSelectedRanges/revertSelectedRanges` currently all do the same thing. They just have different names because different commands need to exist based on where they appear in the UI. Sad, but it's how it is.

---

So... my approach to not having line based operations would be to come up with new commands to:

- Stage selected lines
- Unstage selected lines
- Revert selected lines

This is what #9475 tracks.

**Comment by Git-Lior** [comment]
> `stageChange/revertChange` and `stageSelectedRanges/unstageSelectedRanges/revertSelectedRanges` currently all do the same thing.

@joaomoreno You claim that `stageChange` and `stageSelectedRanges` are the same, but they are implemented differently in the code:

https://github.com/Microsoft/vscode/blob/10cf7b121ac072eaaad3fb1e3411bb3d791e3fa1/extensions/git/src/commands.ts#L831-L862

Specifically, the main difference is that `stageSelectedRange` (as well as `unstageSelectedRange` and with my PR also `revertSelectedRange`) all do this:

https://github.com/Microsoft/vscode/blob/10cf7b121ac072eaaad3fb1e3411bb3d791e3fa1/extensions/git/src/commands.ts#L852-L854

In short, they use `intersectDiffWithRange` to only affect the lines which the user selects currently.

You can rename these commands to `stageSelectedLines` etc. if you want, but they will be the same as `stageSelectedChanges` (probably with changes to `intersectDiffWithRange` which currently isn't perfect).

Am I missing something here?

**Comment by joaomoreno** [comment]
Damn. I completely forgot we had that. Absolutely right! I'll take a good look at this and iron out the naming then.

**Comment by SirIntruder** [comment]
thanks a ton! @Git-Lior :smile: 

I am now looking into code and can't see a reason for the annoying scrolling behaviour. Both revert commands scroll the editor to the document end, but "revert hunk" somehow returns document to where it was before, with a split second delay. Any ideas?

**Comment by joaomoreno** [comment]
@Git-Lior This does not seem to fix it. The behavior in the GIF is not what a user would expect, but instead:

```
test 1
test 2
change 2
test 4
```

This is a tough problem to solve since it seems that it can only be correctly tackled if the original range is the same size as the modified range.
